state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,353 @@
+	fn partial_cmp(&self, other: &Self) -> Option<core::cmp::Ordering> {
+		Some(self.cmp(other))
+	}
+}
+
+impl_pallet_safe_mode!(PalletSafeMode; reconciliation_enabled);
+
+#[frame_support::pallet]
+pub mod pallet {
+	use cf_chains::{dot::PolkadotCrypto, ForeignChain};
+	use cf_primitives::EgressId;
+
+	use super::*;
+	#[pallet::config]
+	#[pallet::disable_frame_system_supertrait_check]
+	pub trait Config: Chainflip {
+		/// Because we want to emit events when there is a config change during
+		/// an runtime upgrade.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Handles egress for all chains.
+		type EgressHandler: EgressApi<AnyChain>;
+
+		/// Polkadot environment.
+		type PolkadotKeyProvider: KeyProvider<PolkadotCrypto>;
+
+		/// Safe mode configuration.
+		type SafeMode: Get<PalletSafeMode>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Refund amount is too low to cover the egress fees. In this case, the refund is skipped.
+		RefundAmountTooLow,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Refund scheduled for a validator.
+		RefundScheduled {
+			egress_id: EgressId,
+			destination: ForeignChainAddress,
+			amount: AssetAmount,
+		},
+		/// The refund was skipped because of the given reason.
+		RefundSkipped { reason: DispatchError },
+		/// The Vault is running a deficit: we owe more than we have set aside for refunds.
+		VaultDeficitDetected {
+			chain: ForeignChain,
+			amount_owed: AssetAmount,
+			available: AssetAmount,
+		},
+	}
+
+	#[pallet::pallet]
+	#[pallet::storage_version(PALLET_VERSION)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Liabilities are funds that are owed to some external party.
+	#[pallet::storage]
+	pub type Liabilities<T: Config> =
+		StorageMap<_, Twox64Concat, Asset, BTreeMap<ExternalOwner, AssetAmount>, ValueQuery>;
+
+	/// Funds that have been set aside to refund external [Liabilities].
+	#[pallet::storage]
+	pub type WithheldAssets<T: Config> =
+		StorageMap<_, Twox64Concat, Asset, AssetAmount, ValueQuery>;
+}
+
+impl<T: Config> Pallet<T> {
+	fn refund_via_egress(
+		chain: ForeignChain,
+		address: ForeignChainAddress,
+		amount: AssetAmount,
+	) -> Result<(), DispatchError> {
+		match T::EgressHandler::schedule_egress(chain.gas_asset(), amount, address.clone(), None)
+			.map_err(Into::into)
+			.and_then(|result @ ScheduledEgressDetails { egress_amount, fee_withheld, .. }| {
+				if egress_amount < REFUND_FEE_MULTIPLE * fee_withheld {
+					Err(Error::<T>::RefundAmountTooLow.into())
+				} else {
+					Ok(result)
+				}
+			}) {
+			Ok(ScheduledEgressDetails { egress_id, .. }) => {
+				Self::deposit_event(Event::RefundScheduled {
+					egress_id,
+					destination: address,
+					amount,
+				});
+				Ok(())
+			},
+			Err(err) => {
+				Self::deposit_event(Event::RefundSkipped { reason: err });
```

Comment:
I don't think this is true right? If we've scheduled it, it's already been scheduled. This error bubbles up to just a break out of a loop. We're not in a transactional block of code here.

Reply:
Ahhh good catch. You mean, we have to do this inside a `with_storage_layer` or `with_transaction` to make this work.

Reply:
I implemented it now, the same way we do it when we burn flip.